<a href="https://colab.research.google.com/github/ginolaratro/LLM-projects/blob/main/neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
import numpy as np

# Fake dataset
# 100 samples, 4 features each
X = np.random.rand(100, 4)

# 3 classes (0, 1, 2)
y = np.random.randint(0, 3, size=(100,))

# Build model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(8, activation='relu', input_shape=(4,)),
    tf.keras.layers.Dense(3)  # <- 3 outputs = 3 logits
])

# Compile
model.compile(
    optimizer='adam',
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=['accuracy']
)

# Train
model.fit(X, y, epochs=10)


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step - accuracy: 0.2763 - loss: 1.1086
Epoch 2/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.3429 - loss: 1.1022 
Epoch 3/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.3209 - loss: 1.0997
Epoch 4/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - accuracy: 0.2863 - loss: 1.1041
Epoch 5/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.3205 - loss: 1.0995
Epoch 6/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.3320 - loss: 1.0980
Epoch 7/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 16ms/step - accuracy: 0.3535 - loss: 1.0996
Epoch 8/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.3378 - loss: 1.0965
Epoch 9/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step - accuracy: 0.3513 - loss: 1.0944
Epoch 10/10
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.3938 - loss: 1.0956


In [ ]:
X

array([[0.59455109, 0.09297815, 0.78743409, 0.08490913],
       [0.82434318, 0.83552866, 0.3682319 , 0.77578502],
       [0.47267006, 0.84834805, 0.55696681, 0.98466905],
       [0.77185967, 0.89500357, 0.04069778, 0.16641197],
       [0.77560737, 0.19604355, 0.82140162, 0.71919577],
       [0.16634876, 0.94514934, 0.73647979, 0.12058142],
       [0.48569275, 0.53951813, 0.40495994, 0.44572868],
       [0.1682396 , 0.08077451, 0.86209302, 0.16174118],
       [0.69208107, 0.6043585 , 0.78623047, 0.36185601],
       [0.54486464, 0.47842145, 0.55151662, 0.70251362],
       [0.58999854, 0.31693625, 0.63276763, 0.88390825],
       [0.88331353, 0.88395552, 0.67680744, 0.74079422],
       [0.58756076, 0.58073074, 0.67742973, 0.06449461],
       [0.45954617, 0.6626138 , 0.07750123, 0.86144959],
       [0.99794706, 0.19059556, 0.5885372 , 0.99939481],
       [0.59284585, 0.2556106 , 0.45029435, 0.69505588],
       [0.32663461, 0.49227349, 0.45669221, 0.57123094],
       [0.82603208, 0.97828315,

In [ ]:
import tensorflow as tf
import numpy as np

# Load real handwritten digit data
(X_train, y_train), (X_test, y_test) = tf.keras.datasets.mnist.load_data()


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [ ]:
X_train = X_train / 255.0
X_test = X_test / 255.0

model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),  # turn image into 784 numbers
    tf.keras.layers.Dense(128, activation='relu'),  # learn patterns
    tf.keras.layers.Dense(10)  # 10 logits (for digits 0–9)
])


model.compile(
    optimizer='adam',
    loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=['accuracy']
)

model.fit(X_train, y_train, epochs=5)



Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.1078 - loss: 2.3018
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.1122 - loss: 2.3013
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.1133 - loss: 2.3014
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.1140 - loss: 2.3011
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.1119 - loss: 2.3013


In [ ]:
model.evaluate(X_test, y_test)


313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1160 - loss: 2.3013


[2.301264762878418, 0.11349999904632568]

In [ ]:
sample = X_test[0:1]

logits = model(sample)
print("Logits:", logits.numpy())


Logits: [[ 0.00197627  0.11738122 -0.03177003  0.0131141  -0.030847   -0.08435042
  -0.00325851  0.03813067 -0.0354943  -0.02360712]]


In [ ]:
probs = tf.nn.softmax(logits)

print("Probabilities:", probs.numpy())
print("Predicted digit:", np.argmax(probs))
print("Actual digit:", y_test[0])


Probabilities: [[0.10045296 0.11274116 0.09711961 0.10157804 0.09720929 0.09214496
  0.09992848 0.10415123 0.09675858 0.09791563]]
Predicted digit: 1
Actual digit: 7


Logits are the raw, unnormalized output scores (ranging from \(-\infty \) to \(+\infty \)) from the final layer of a neural network before an activation function (like sigmoid or softmax) is applied. They represent the model's initial prediction confidence for classification tasks, where higher values indicate a higher likelihood for a class

Next we will implement Cross-Entropy Loss.
This will tell us how accurate our results are

> Add blockquote




In [2]:
import tensorflow as tf
import numpy as np # Added for consistency, though not strictly needed for this specific loss calculation here.

# The previous execution failed because 'y_test' and 'logits' were not defined in the current kernel session.
# To demonstrate the cross-entropy loss calculation, we will re-initialize these variables
# using the values that were successfully computed and displayed in previous cells.

# Re-defining logits with the value obtained from `mi5jbXaLrecH`
logits = tf.constant([[ 0.00197627,  0.11738122, -0.03177003,  0.0131141,  -0.030847,   -0.08435042,
                      -0.00325851,  0.03813067, -0.0354943,  -0.02360712]], dtype=tf.float32)

# Re-defining y_test[0] with the actual digit value obtained from `9WpXoaxkrlod`
actual_digit = 7
y_test_single_sample = tf.constant([actual_digit], dtype=tf.int32) # Needs to be a tensor with batch dimension

# Calculate Sparse Categorical Crossentropy Loss for the sample
loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

# Calculate the loss for the sample
sample_loss = loss_object(y_test_single_sample, logits)

print(f"Logits for the sample: {logits.numpy().flatten()}")
print(f"Actual digit for the sample: {actual_digit}")
print(f"Calculated Sparse Categorical Crossentropy Loss: {sample_loss.numpy()}")

Logits for the sample: [ 0.00197627  0.11738122 -0.03177003  0.0131141  -0.030847   -0.08435042
 -0.00325851  0.03813067 -0.0354943  -0.02360712]
Actual digit for the sample: 7
Calculated Sparse Categorical Crossentropy Loss: 2.261911153793335


# Task
Implement Positional Encoding for a Transformer model.

## Understand Transformer Architecture

### Subtask:
Start by understanding the key components of a transformer model, including self-attention, multi-head attention, positional encoding, and the encoder-decoder structure. This will inform the implementation.


### Understanding the Transformer Architecture

The Transformer model, introduced in the paper "Attention Is All You Need," revolutionized sequence-to-sequence tasks, particularly in natural language processing. Unlike recurrent neural networks (RNNs) and convolutional neural networks (CNNs), it relies entirely on attention mechanisms to draw global dependencies between input and output.

#### Key Components:

1.  **Self-Attention (Scaled Dot-Product Attention):**
    *   This mechanism allows the model to weigh the importance of different words in the input sequence when encoding a single word. It calculates a 'score' for each word relative to every other word, indicating how much attention to pay to other parts of the sequence.
    *   It uses three learnable matrices: Query (Q), Key (K), and Value (V).
    *   The attention score is calculated as: `Attention(Q, K, V) = softmax(Q * K_transpose / sqrt(d_k)) * V`.

2.  **Multi-Head Attention:**
    *   Instead of performing a single attention function, multi-head attention runs `h` different attention mechanisms in parallel. Each 'head' learns different aspects of the relationships between words.
    *   The outputs from these `h` attention heads are concatenated and then linearly transformed into the desired dimension.
    *   This allows the model to jointly attend to information from different representation subspaces at different positions.

3.  **Positional Encoding:**
    *   Since the Transformer architecture does not inherently process sequences in order (due to the lack of recurrence or convolution), positional encoding is added to the input embeddings to inject information about the relative or absolute position of tokens in the sequence.
    *   These encodings are typically fixed (not learned) sinusoidal functions of different frequencies.

4.  **Feed-Forward Networks (FFN):**
    *   Each position in the Transformer's encoder and decoder contains a fully connected feed-forward network, which is applied identically and independently to each position.
    *   This typically consists of two linear transformations with a ReLU activation in between: `FFN(x) = max(0, x * W1 + b1) * W2 + b2`.

5.  **Residual Connections & Layer Normalization:**
    *   **Residual Connections:** Each sub-layer (self-attention, feed-forward) in the encoder and decoder is surrounded by a residual connection (`x + Sublayer(x)`), which helps in training deeper networks by allowing gradients to flow through bypass paths.
    *   **Layer Normalization:** After the residual connection, layer normalization is applied. Unlike batch normalization, layer normalization normalizes the inputs across the features for each sample independently, rather than across a batch.

6.  **Encoder-Decoder Structure:**
    *   **Encoder:** Composed of a stack of `N` identical layers. Each layer has two sub-layers: a multi-head self-attention mechanism and a position-wise fully connected feed-forward network. The encoder processes the input sequence and produces a sequence of continuous representations.
    *   **Decoder:** Also composed of a stack of `N` identical layers. Each layer has three sub-layers: a masked multi-head self-attention mechanism (to prevent attending to future positions during training), a multi-head attention layer that attends to the output of the encoder, and a position-wise fully connected feed-forward network. The decoder generates the output sequence one token at a time, taking the encoder's output and the previously generated tokens as input.

## Implement Positional Encoding

### Subtask:
Implement the positional encoding mechanism that adds information about the position of tokens in the sequence, as transformers do not inherently process sequences in order.


**Reasoning**:
The first step is to import the necessary libraries, `tensorflow` and `numpy`, as specified in the instructions to prepare for defining the positional encoding components.



In [1]:
import tensorflow as tf
import numpy as np

print("TensorFlow and NumPy imported successfully.")

TensorFlow and NumPy imported successfully.


**Reasoning**:
Following the instructions, the next step is to define the `get_angles` function which calculates the arguments for the sine and cosine functions, as required for positional encoding.



In [2]:
def get_angles(pos, i, d_model):
    angle_rates = 1 / np.power(10000, (2 * (i // 2)) / np.float32(d_model))
    return pos * angle_rates

print("get_angles function defined.")

get_angles function defined.


**Reasoning**:
As per the instructions, the next step is to create the `PositionalEncoding` class, which will generate the positional encodings using the previously defined `get_angles` function and the sine/cosine formulas.



In [3]:
class PositionalEncoding(tf.keras.layers.Layer):
    def __init__(self, position, d_model):
        super().__init__()
        self.pos_encoding = self.positional_encoding(position, d_model)

    def get_angles(self, pos, i, d_model):
        angle_rates = 1 / np.power(10000, (2 * (i // 2)) / np.float32(d_model))
        return pos * angle_rates

    def positional_encoding(self, position, d_model):
        angle_rads = self.get_angles(np.arange(position)[:, np.newaxis],
                                     np.arange(d_model)[np.newaxis, :],
                                     d_model)

        # apply sin to even indices in the array; 2i
        angle_rads[:, 0::2] = np.sin(angle_rads[:, 0::2])

        # apply cos to odd indices in the array; 2i + 1
        angle_rads[:, 1::2] = np.cos(angle_rads[:, 1::2])

        pos_encoding = angle_rads[np.newaxis, ...]
        return tf.cast(pos_encoding, dtype=tf.float32)

    def call(self, inputs):
        return inputs + self.pos_encoding[:, :tf.shape(inputs)[1], :]

print("PositionalEncoding class defined.")

PositionalEncoding class defined.


**Reasoning**:
To complete the subtask, the next step is to test the `PositionalEncoding` class by creating an instance with specific `position` (max_sequence_length) and `d_model` values, and then printing the shape of the generated positional encodings to verify the implementation.



In [4]:
max_sequence_length = 50
d_model = 512

# Create an instance of PositionalEncoding
pos_encoding_layer = PositionalEncoding(max_sequence_length, d_model)

# Access the generated positional encodings
# The pos_encoding attribute already has the batch dimension (1, max_sequence_length, d_model)
positional_encodings = pos_encoding_layer.pos_encoding

print(f"Shape of generated positional encodings: {positional_encodings.shape}")

# You can also test the call method by providing dummy input
dummy_inputs = tf.zeros((1, 10, d_model), dtype=tf.float32) # Batch size 1, sequence length 10
output_with_pe = pos_encoding_layer(dummy_inputs)
print(f"Shape of output after adding positional encodings (dummy input): {output_with_pe.shape}")

Shape of generated positional encodings: (1, 50, 512)
Shape of output after adding positional encodings (dummy input): (1, 10, 512)


## Prepare Data for Transformer

### Subtask:
Prepare dummy data including tokenization, numericalization, padding, and batching to simulate a sequence-to-sequence task for demonstrating Transformer components.


**Reasoning**:
To prepare dummy data for the Transformer, the first step is to define simple integer-based dummy source and target vocabularies, which will be used for numericalization, and then create dummy input and target sequences based on these vocabularies.



In [5]:
import tensorflow as tf
import numpy as np

# 1. Define dummy source and target vocabularies
source_vocab = {
    0: '<pad>', 1: '<start>', 2: '<end>',
    3: 'hello', 4: 'world', 5: 'how',
    6: 'are', 7: 'you', 8: 'i',
    9: 'am', 10: 'fine'
}
target_vocab = {
    0: '<pad>', 1: '<start>', 2: '<end>',
    3: 'bonjour', 4: 'monde', 5: 'comment',
    6: 'allez', 7: 'vous', 8: 'je',
    9: 'vais', 10: 'bien'
}

# Create reverse vocabularies for easy lookup (e.g., for numericalization)
# In a real scenario, you'd typically have a tokenizer that maps text to IDs
source_id_to_word = {v: k for k, v in source_vocab.items()}
target_id_to_word = {v: k for k, v in target_vocab.items()}

# 2. Create dummy input and target sequences (numericalized)
# Sequences are represented as lists of integers (token IDs)
# Adding <start> and <end> tokens for target sequences usually
# Input: "hello world how are you"
# Target: "bonjour monde comment allez vous"
input_sequences_raw = [
    [source_id_to_word['hello'], source_id_to_word['world'], source_id_to_word['how'], source_id_to_word['are'], source_id_to_word['you']],
    [source_id_to_word['i'], source_id_to_word['am'], source_id_to_word['fine']],
    [source_id_to_word['how'], source_id_to_word['are'], source_id_to_word['you'], source_id_to_word['hello']]
]
target_sequences_raw = [
    [target_id_to_word['<start>'], target_id_to_word['bonjour'], target_id_to_word['monde'], target_id_to_word['comment'], target_id_to_word['allez'], target_id_to_word['vous'], target_id_to_word['<end>']],
    [target_id_to_word['<start>'], target_id_to_word['je'], target_id_to_word['vais'], target_id_to_word['bien'], target_id_to_word['<end>']],
    [target_id_to_word['<start>'], target_id_to_word['comment'], target_id_to_word['allez'], target_id_to_word['vous'], target_id_to_word['bonjour'], target_id_to_word['<end>']]
]

print("Dummy vocabularies and raw sequences created.")

Dummy vocabularies and raw sequences created.


**Reasoning**:
Now that the dummy vocabularies and raw sequences are defined, the next logical step is to determine a maximum sequence length, pad these raw sequences to that length using the padding token, and then group them into TensorFlow tensor batches for further processing, as per the instructions.



In [6]:
PAD_ID = 0

# Determine max sequence length for padding (can be adjusted based on needs)
max_input_seq_len = max(len(seq) for seq in input_sequences_raw)
max_target_seq_len = max(len(seq) for seq in target_sequences_raw)

# 3. Pad the dummy input and target sequences
# Using tf.keras.preprocessing.sequence.pad_sequences for convenience
padded_input_sequences = tf.keras.preprocessing.sequence.pad_sequences(
    input_sequences_raw, maxlen=max_input_seq_len, padding='post', value=PAD_ID
)
padded_target_sequences = tf.keras.preprocessing.sequence.pad_sequences(
    target_sequences_raw, maxlen=max_target_seq_len, padding='post', value=PAD_ID
)

# 4. Group the padded sequences into dummy batches (convert to tf.Tensor)
# For simplicity, we'll use all sequences as a single batch
input_batch = tf.constant(padded_input_sequences, dtype=tf.int32)
target_batch = tf.constant(padded_target_sequences, dtype=tf.int32)

# 5. Print the shapes and a sample to verify
print(f"Shape of input_batch: {input_batch.shape}")
print(f"Shape of target_batch: {target_batch.shape}")
print(f"\nSample input sequence (numerical): {input_batch[0].numpy()}")
print(f"Sample target sequence (numerical): {target_batch[0].numpy()}")

# Optionally, decode the sample to words for better understanding
def decode_sequence(sequence, vocab):
    return ' '.join([vocab[idx] for idx in sequence if idx != PAD_ID])

print(f"\nSample input sequence (decoded): {decode_sequence(input_batch[0].numpy(), source_vocab)}")
print(f"Sample target sequence (decoded): {decode_sequence(target_batch[0].numpy(), target_vocab)}")


Shape of input_batch: (3, 5)
Shape of target_batch: (3, 7)

Sample input sequence (numerical): [3 4 5 6 7]
Sample target sequence (numerical): [1 3 4 5 6 7 2]

Sample input sequence (decoded): hello world how are you
Sample target sequence (decoded): <start> bonjour monde comment allez vous <end>


## Build Multi-Head Attention Mechanism

### Subtask:
Develop the multi-head attention layer, which allows the model to jointly attend to information from different representation subspaces at different positions.


**Reasoning**:
As instructed, the first step in developing the `MultiHeadAttention` layer is to define the `MultiHeadAttention` class inheriting from `tf.keras.layers.Layer`, and within its `__init__` method, initialize `d_model`, `num_heads`, calculate `depth`, and set up the necessary `tf.keras.layers.Dense` layers for queries, keys, values, and the final output.



In [7]:
class MultiHeadAttention(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads):
        super(MultiHeadAttention, self).__init__()
        self.num_heads = num_heads
        self.d_model = d_model

        assert d_model % self.num_heads == 0

        self.depth = d_model // self.num_heads

        self.wq = tf.keras.layers.Dense(d_model)
        self.wk = tf.keras.layers.Dense(d_model)
        self.wv = tf.keras.layers.Dense(d_model)

        self.dense = tf.keras.layers.Dense(d_model)

        print(f"MultiHeadAttention initialized with d_model={d_model}, num_heads={num_heads}, depth={self.depth}")

print("MultiHeadAttention class definition started with __init__.")

MultiHeadAttention class definition started with __init__.


**Reasoning**:
As instructed, the next step is to implement the `split_heads` method within the `MultiHeadAttention` class, which will reshape the input tensor to prepare it for multi-head attention.



In [8]:
class MultiHeadAttention(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads):
        super(MultiHeadAttention, self).__init__()
        self.num_heads = num_heads
        self.d_model = d_model

        assert d_model % self.num_heads == 0

        self.depth = d_model // self.num_heads

        self.wq = tf.keras.layers.Dense(d_model)
        self.wk = tf.keras.layers.Dense(d_model)
        self.wv = tf.keras.layers.Dense(d_model)

        self.dense = tf.keras.layers.Dense(d_model)

        print(f"MultiHeadAttention initialized with d_model={d_model}, num_heads={num_heads}, depth={self.depth}")

    def split_heads(self, x, batch_size):
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.depth))
        return tf.transpose(x, perm=[0, 2, 1, 3])

print("MultiHeadAttention class updated with split_heads method.")

MultiHeadAttention class updated with split_heads method.


**Reasoning**:
As per the instructions, the next step is to implement the `scaled_dot_product_attention` function, which performs the core attention calculation.



In [9]:
def scaled_dot_product_attention(q, k, v, mask):
    matmul_qk = tf.matmul(q, k, transpose_b=True)  # (..., seq_len_q, seq_len_k)

    dk = tf.cast(tf.shape(k)[-1], tf.float32)
    scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)

    if mask is not None:
        scaled_attention_logits += (mask * -1e9)

    attention_weights = tf.nn.softmax(scaled_attention_logits, axis=-1)  # (..., seq_len_q, seq_len_k)

    output = tf.matmul(attention_weights, v)  # (..., seq_len_q, depth_v)

    return output, attention_weights

print("scaled_dot_product_attention function defined.")

scaled_dot_product_attention function defined.


**Reasoning**:
As instructed, the next step is to implement the `call` method within the `MultiHeadAttention` class, which will orchestrate the dense transformations, splitting heads, performing scaled dot-product attention, and concatenating results.



In [10]:
class MultiHeadAttention(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads):
        super(MultiHeadAttention, self).__init__()
        self.num_heads = num_heads
        self.d_model = d_model

        assert d_model % self.num_heads == 0

        self.depth = d_model // self.num_heads

        self.wq = tf.keras.layers.Dense(d_model)
        self.wk = tf.keras.layers.Dense(d_model)
        self.wv = tf.keras.layers.Dense(d_model)

        self.dense = tf.keras.layers.Dense(d_model)

    def split_heads(self, x, batch_size):
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.depth))
        return tf.transpose(x, perm=[0, 2, 1, 3])

    def call(self, v, k, q, mask):
        batch_size = tf.shape(q)[0]

        q = self.wq(q)  # (batch_size, seq_len, d_model)
        k = self.wk(k)  # (batch_size, seq_len, d_model)
        v = self.wv(v)  # (batch_size, seq_len, d_model)

        q = self.split_heads(q, batch_size)  # (batch_size, num_heads, seq_len_q, depth)
        k = self.split_heads(k, batch_size)  # (batch_size, num_heads, seq_len_k, depth)
        v = self.split_heads(v, batch_size)  # (batch_size, num_heads, seq_len_v, depth)

        # scaled_attention.shape == (batch_size, num_heads, seq_len_q, depth)
        # attention_weights.shape == (batch_size, num_heads, seq_len_q, seq_len_k)
        scaled_attention, attention_weights = scaled_dot_product_attention(
            q, k, v, mask)

        scaled_attention = tf.transpose(scaled_attention, perm=[0, 2, 1, 3])  # (batch_size, seq_len_q, num_heads, depth)

        concat_attention = tf.reshape(scaled_attention,
                                      (batch_size, -1, self.d_model))  # (batch_size, seq_len_q, d_model)

        output = self.dense(concat_attention)  # (batch_size, seq_len_q, d_model)

        return output, attention_weights

print("MultiHeadAttention class updated with call method.")

MultiHeadAttention class updated with call method.


**Reasoning**:
As instructed, the final step for this subtask is to test the `MultiHeadAttention` layer by creating an instance with arbitrary `d_model` and `num_heads`, and then passing dummy input tensors to its `call` method to verify the shapes of the output and attention weights.



In [11]:
num_heads = 8
d_model = 512
batch_size = 64
seq_len = 100 # Example sequence length

# Create an instance of MultiHeadAttention
mha = MultiHeadAttention(d_model, num_heads)

# Create dummy input tensors for query, key, and value
# In self-attention, Q, K, V are typically the same input sequence
dummy_input = tf.random.uniform((batch_size, seq_len, d_model))

# Assuming no mask for this initial test
output, attn_weights = mha(v=dummy_input, k=dummy_input, q=dummy_input, mask=None)

print(f"MultiHeadAttention output shape: {output.shape}") # Expected: (batch_size, seq_len, d_model)
print(f"Attention weights shape: {attn_weights.shape}") # Expected: (batch_size, num_heads, seq_len, seq_len)

MultiHeadAttention output shape: (64, 100, 512)
Attention weights shape: (64, 8, 100, 100)


## Construct Transformer Encoder Layer

### Subtask:
Implement a single transformer encoder layer, consisting of a multi-head self-attention mechanism and a position-wise feed-forward network.


**Reasoning**:
As instructed, the first step to construct the Transformer Encoder Layer is to define the `point_wise_feed_forward_network` helper function, which will be used within the `EncoderLayer` to perform position-wise transformations.



In [12]:
def point_wise_feed_forward_network(d_model, dff):
    return tf.keras.Sequential([
        tf.keras.layers.Dense(dff, activation='relu'),  # (batch_size, seq_len, dff)
        tf.keras.layers.Dense(d_model)  # (batch_size, seq_len, d_model)
    ])

print("point_wise_feed_forward_network function defined.")

point_wise_feed_forward_network function defined.


**Reasoning**:
As instructed, the next step is to define the `EncoderLayer` class inheriting from `tf.keras.layers.Layer` and implement its `__init__` method, initializing `d_model`, `num_heads`, `MultiHeadAttention`, `point_wise_feed_forward_network`, `LayerNormalization` and `Dropout` layers.



In [13]:
class EncoderLayer(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads, dff, rate=0.1):
        super(EncoderLayer, self).__init__()

        self.mha = MultiHeadAttention(d_model, num_heads)
        self.ffn = point_wise_feed_forward_network(d_model, dff)

        self.layernorm1 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

        self.dropout1 = tf.keras.layers.Dropout(rate)
        self.dropout2 = tf.keras.layers.Dropout(rate)

        print(f"EncoderLayer initialized with d_model={d_model}, num_heads={num_heads}, dff={dff}, dropout_rate={rate}")

print("EncoderLayer class definition started with __init__.")

EncoderLayer class definition started with __init__.


**Reasoning**:
As instructed, the next step is to implement the `call` method within the `EncoderLayer` class, which will orchestrate the multi-head self-attention, feed-forward network, residual connections, layer normalization, and dropout.



In [14]:
class EncoderLayer(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads, dff, rate=0.1):
        super(EncoderLayer, self).__init__()

        self.mha = MultiHeadAttention(d_model, num_heads)
        self.ffn = point_wise_feed_forward_network(d_model, dff)

        self.layernorm1 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

        self.dropout1 = tf.keras.layers.Dropout(rate)
        self.dropout2 = tf.keras.layers.Dropout(rate)

    def call(self, x, training, mask):
        # Multi-head attention (self-attention)
        attn_output, _ = self.mha(x, x, x, mask)  # (batch_size, input_seq_len, d_model)
        attn_output = self.dropout1(attn_output, training=training)
        out1 = self.layernorm1(x + attn_output)  # (batch_size, input_seq_len, d_model)

        # Feed Forward Network
        ffn_output = self.ffn(out1)  # (batch_size, input_seq_len, d_model)
        ffn_output = self.dropout2(ffn_output, training=training)
        out2 = self.layernorm2(out1 + ffn_output)  # (batch_size, input_seq_len, d_model)

        return out2

print("EncoderLayer class updated with call method.")

EncoderLayer class updated with call method.


**Reasoning**:
As instructed, the final step for this subtask is to test the `EncoderLayer` by creating an instance with arbitrary parameters and passing dummy input tensors to its `call` method to verify the output shape.



In [15]:
num_heads = 8
d_model = 512
dff = 2048 # Dimension of the feed-forward network inner layer
dropout_rate = 0.1
batch_size = 64
seq_len = 100 # Example sequence length

# Create an instance of EncoderLayer
encoder_layer = EncoderLayer(d_model, num_heads, dff, dropout_rate)

# Create dummy input tensor
dummy_input_encoder = tf.random.uniform((batch_size, seq_len, d_model))

# Call the encoder layer with dummy input (no mask, training=True for dropout effect)
encoder_output = encoder_layer(dummy_input_encoder, training=True, mask=None)

print(f"EncoderLayer output shape: {encoder_output.shape}") # Expected: (batch_size, seq_len, d_model)


EncoderLayer output shape: (64, 100, 512)


## Construct Transformer Decoder Layer

### Subtask:
Implement a single transformer decoder layer, which includes masked multi-head self-attention, multi-head encoder-decoder attention, and a position-wise feed-forward network.


**Reasoning**:
As instructed, the first step to construct the Transformer Decoder Layer is to define the `DecoderLayer` class inheriting from `tf.keras.layers.Layer` and implement its `__init__` method, initializing all the necessary components such as two `MultiHeadAttention` layers, a `point_wise_feed_forward_network`, three `LayerNormalization` layers, and three `Dropout` layers.



In [16]:
class DecoderLayer(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads, dff, rate=0.1):
        super(DecoderLayer, self).__init__()

        self.mha1 = MultiHeadAttention(d_model, num_heads)  # Masked self-attention
        self.mha2 = MultiHeadAttention(d_model, num_heads)  # Encoder-decoder attention

        self.ffn = point_wise_feed_forward_network(d_model, dff)

        self.layernorm1 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm3 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

        self.dropout1 = tf.keras.layers.Dropout(rate)
        self.dropout2 = tf.keras.layers.Dropout(rate)
        self.dropout3 = tf.keras.layers.Dropout(rate)

        print(f"DecoderLayer initialized with d_model={d_model}, num_heads={num_heads}, dff={dff}, dropout_rate={rate}")

print("DecoderLayer class definition started with __init__.")

DecoderLayer class definition started with __init__.


**Reasoning**:
As instructed, the next step is to implement the `call` method within the `DecoderLayer` class, which will orchestrate the masked multi-head self-attention, multi-head encoder-decoder attention, feed-forward network, residual connections, layer normalization, and dropout.



In [17]:
class DecoderLayer(tf.keras.layers.Layer):
    def __init__(self, d_model, num_heads, dff, rate=0.1):
        super(DecoderLayer, self).__init__()

        self.mha1 = MultiHeadAttention(d_model, num_heads)  # Masked self-attention
        self.mha2 = MultiHeadAttention(d_model, num_heads)  # Encoder-decoder attention

        self.ffn = point_wise_feed_forward_network(d_model, dff)

        self.layernorm1 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm2 = tf.keras.layers.LayerNormalization(epsilon=1e-6)
        self.layernorm3 = tf.keras.layers.LayerNormalization(epsilon=1e-6)

        self.dropout1 = tf.keras.layers.Dropout(rate)
        self.dropout2 = tf.keras.layers.Dropout(rate)
        self.dropout3 = tf.keras.layers.Dropout(rate)

    def call(self, x, enc_output, training, look_ahead_mask, padding_mask):
        # Masked Multi-head self-attention
        attn1, attn_weights_block1 = self.mha1(x, x, x, look_ahead_mask)
        attn1 = self.dropout1(attn1, training=training)
        out1 = self.layernorm1(attn1 + x) # (batch_size, target_seq_len, d_model)

        # Multi-head encoder-decoder attention
        # q = out1, k = enc_output, v = enc_output
        attn2, attn_weights_block2 = self.mha2(
            enc_output, enc_output, out1, padding_mask)
        attn2 = self.dropout2(attn2, training=training)
        out2 = self.layernorm2(attn2 + out1) # (batch_size, target_seq_len, d_model)

        # Feed Forward Network
        ffn_output = self.ffn(out2) # (batch_size, target_seq_len, d_model)
        ffn_output = self.dropout3(ffn_output, training=training)
        out3 = self.layernorm3(ffn_output + out2) # (batch_size, target_seq_len, d_model)

        return out3, attn_weights_block1, attn_weights_block2

print("DecoderLayer class updated with call method.")


DecoderLayer class updated with call method.


## Assemble Full Transformer Model

### Subtask:
Implement the Encoder part of the Transformer model by combining input embeddings, positional encoding, and multiple EncoderLayer instances.


**Reasoning**:
As instructed, the first step is to define the `Encoder` class inheriting from `tf.keras.layers.Layer` and implement its `__init__` method, initializing the embedding layer, positional encoding layer, a list of encoder layers, and a dropout layer.



In [18]:
class Encoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Encoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(input_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, self.d_model)

        self.enc_layers = [
            EncoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]

        self.dropout = tf.keras.layers.Dropout(rate)

        print(f"Encoder initialized with num_layers={num_layers}, d_model={d_model}, num_heads={num_heads}, dff={dff}, input_vocab_size={input_vocab_size}, maximum_position_encoding={maximum_position_encoding}, dropout_rate={rate}")

print("Encoder class definition started with __init__.")

Encoder class definition started with __init__.


**Reasoning**:
As instructed, the next step is to implement the `call` method within the `Encoder` class, which will apply input embeddings, add positional encoding, apply dropout, and pass the data through multiple `EncoderLayer` instances.



In [19]:
class Encoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Encoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(input_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, self.d_model)

        self.enc_layers = [
            EncoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]

        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, training, mask):
        seq_len = tf.shape(x)[1]

        x = self.embedding(x)  # (batch_size, input_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        x += self.pos_encoding[:, :seq_len, :]

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x = self.enc_layers[i](x, training, mask)

        return x  # (batch_size, input_seq_len, d_model)

print("Encoder class updated with call method.")

Encoder class updated with call method.


## Assemble Full Transformer Model

### Subtask:
Implement the Decoder part of the Transformer model by combining input embeddings, positional encoding, and multiple DecoderLayer instances.


**Reasoning**:
As instructed, the next step is to define the `Decoder` class, inheriting from `tf.keras.layers.Layer`, and implement its `__init__` method to initialize the embedding layer, positional encoding layer, multiple `DecoderLayer` instances, and a dropout layer.



In [20]:
class Decoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, target_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Decoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(target_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, d_model)

        self.dec_layers = [
            DecoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]
        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, enc_output, training, look_ahead_mask, padding_mask):
        seq_len = tf.shape(x)[1]
        attention_weights = {}

        x = self.embedding(x)  # (batch_size, target_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        x += self.pos_encoding[:, :seq_len, :]

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x, block1, block2 = self.dec_layers[i](x, enc_output, training,
                                                    look_ahead_mask, padding_mask)

            attention_weights[f'decoder_layer{i+1}_block1'] = block1
            attention_weights[f'decoder_layer{i+1}_block2'] = block2

        # x.shape == (batch_size, target_seq_len, d_model)
        return x, attention_weights

print("Decoder class definition completed with __init__ and call methods.")

Decoder class definition completed with __init__ and call methods.


## Assemble Full Transformer Model

### Subtask:
Assemble the complete Transformer model by combining the encoder and decoder layers, along with input and output embeddings and linear transformations. This includes defining masking functions.


**Reasoning**:
As instructed, the first step to assemble the complete Transformer model is to define the `create_padding_mask` function, which generates a mask for padding tokens.



In [21]:
def create_padding_mask(seq):
    seq = tf.cast(tf.math.equal(seq, PAD_ID), tf.float32)
    # add extra dimensions to add the padding
    # to the attention logits. For example, for a sequence
    # of length 3, the mask will be [[0, 0, 0], [0, 0, 0], [0, 0, 0]] if no padding
    # or [[0, 0, 1], [0, 0, 1], [0, 0, 1]] if the last token is padding.
    return seq[:, tf.newaxis, tf.newaxis, :]

print("create_padding_mask function defined.")

create_padding_mask function defined.


**Reasoning**:
As instructed, the next step is to define the `create_look_ahead_mask` function, which creates a mask to hide future tokens in the decoder's self-attention mechanism.



In [22]:
def create_look_ahead_mask(size):
    mask = 1 - tf.linalg.band_part(tf.ones((size, size)), -1, 0)
    return mask  # (seq_len, seq_len)

print("create_look_ahead_mask function defined.")

create_look_ahead_mask function defined.


**Reasoning**:
As instructed, the next step is to define the `Transformer` class, inheriting from `tf.keras.Model`. In its `__init__` method, I will initialize the `Encoder`, `Decoder`, and a final `tf.keras.layers.Dense` layer for the output vocabulary.



In [23]:
class Transformer(tf.keras.Model):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 target_vocab_size, pe_input, pe_target, rate=0.1):
        super().__init__()
        self.encoder = Encoder(num_layers, d_model, num_heads, dff,
                               input_vocab_size, pe_input, rate)

        self.decoder = Decoder(num_layers, d_model, num_heads, dff,
                               target_vocab_size, pe_target, rate)

        self.final_layer = tf.keras.layers.Dense(target_vocab_size)

        print(f"Transformer initialized with num_layers={num_layers}, d_model={d_model}, num_heads={num_heads}, dff={dff}, input_vocab_size={input_vocab_size}, target_vocab_size={target_vocab_size}, pe_input={pe_input}, pe_target={pe_target}, dropout_rate={rate}")

print("Transformer class definition started with __init__.")

Transformer class definition started with __init__.


**Reasoning**:
As instructed, the next step is to implement the `call` method within the `Transformer` class. This method will orchestrate the data flow through the encoder and decoder, and then apply the final dense layer to produce the output logits.



In [24]:
class Transformer(tf.keras.Model):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 target_vocab_size, pe_input, pe_target, rate=0.1):
        super().__init__()
        self.encoder = Encoder(num_layers, d_model, num_heads, dff,
                               input_vocab_size, pe_input, rate)

        self.decoder = Decoder(num_layers, d_model, num_heads, dff,
                               target_vocab_size, pe_target, rate)

        self.final_layer = tf.keras.layers.Dense(target_vocab_size)

    def call(self, inp, tar, training, enc_padding_mask, look_ahead_mask, dec_padding_mask):
        enc_output = self.encoder(inp, training, enc_padding_mask)  # (batch_size, inp_seq_len, d_model)

        # dec_output.shape == (batch_size, tar_seq_len, d_model)
        dec_output, attention_weights = self.decoder(
            tar, enc_output, training, look_ahead_mask, dec_padding_mask)

        final_output = self.final_layer(dec_output)  # (batch_size, tar_seq_len, target_vocab_size)

        return final_output, attention_weights

print("Transformer class updated with call method.")

Transformer class updated with call method.


**Reasoning**:
As instructed, the final step for this subtask is to test the complete `Transformer` model by creating an instance with appropriate parameters and passing dummy input and target sequences to its `call` method. This will verify the output shape of the logits and the attention weights.



In [25]:
num_layers = 2
d_model = 512
num_heads = 8
dff = 2048

# Use the dummy vocab sizes and max sequence lengths from previous data preparation
input_vocab_size = len(source_vocab)
target_vocab_size = len(target_vocab)
pe_input = max_input_seq_len + 10 # Add some buffer for positional encoding
pe_target = max_target_seq_len + 10 # Add some buffer for positional encoding

# Create an instance of the Transformer model
transformer = Transformer(
    num_layers,
    d_model,
    num_heads,
    dff,
    input_vocab_size,
    target_vocab_size,
    pe_input,
    pe_target,
    rate=0.1
)

# Prepare dummy input and target sequences (using previously created batches)
# input_batch shape: (batch_size, inp_seq_len)
# target_batch shape: (batch_size, tar_seq_len)
dummy_input = input_batch # (3, 5)
dummy_target = target_batch[:, :-1] # Decoder input needs to be shifted right (remove last token)

# Generate masks
enc_padding_mask = create_padding_mask(dummy_input)
dec_padding_mask = create_padding_mask(dummy_input) # Used for encoder-decoder attention
look_ahead_mask = create_look_ahead_mask(tf.shape(dummy_target)[1])
combined_mask = tf.maximum(create_padding_mask(dummy_target), look_ahead_mask) # For decoder self-attention

# Call the transformer model
# Note: training=True to enable dropout
final_output, attention_weights = transformer(
    dummy_input,
    dummy_target,
    training=True,
    enc_padding_mask=enc_padding_mask,
    look_ahead_mask=combined_mask,
    dec_padding_mask=dec_padding_mask
)

print(f"Transformer output (logits) shape: {final_output.shape}") # Expected: (batch_size, tar_seq_len, target_vocab_size)
# Example of one attention weight shape from the decoder
if 'decoder_layer1_block1' in attention_weights:
    print(f"Sample attention weights shape (decoder_layer1_block1): {attention_weights['decoder_layer1_block1'].shape}")


ValueError: Exception encountered when calling Transformer.call().

Only input tensors may be passed as positional arguments. The following argument value should be passed as a keyword argument: True (of type <class 'bool'>)

Arguments received by Transformer.call():
  • inp=tf.Tensor(shape=(3, 5), dtype=int32)
  • tar=tf.Tensor(shape=(3, 6), dtype=int32)
  • training=True
  • enc_padding_mask=tf.Tensor(shape=(3, 1, 1, 5), dtype=float32)
  • look_ahead_mask=tf.Tensor(shape=(3, 1, 6, 6), dtype=float32)
  • dec_padding_mask=tf.Tensor(shape=(3, 1, 1, 5), dtype=float32)

**Reasoning**:
The error indicates that the `training` argument (a boolean) was passed as a positional argument to the `Encoder` and `Decoder` within the `Transformer`'s `call` method, but Keras layers expect such non-tensor arguments to be passed as keyword arguments. I will modify the `call` method in the `Transformer` class to pass `training` and `mask` arguments as keyword arguments to `self.encoder` and `self.decoder`.



In [26]:
class Transformer(tf.keras.Model):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 target_vocab_size, pe_input, pe_target, rate=0.1):
        super().__init__()
        self.encoder = Encoder(num_layers, d_model, num_heads, dff,
                               input_vocab_size, pe_input, rate)

        self.decoder = Decoder(num_layers, d_model, num_heads, dff,
                               target_vocab_size, pe_target, rate)

        self.final_layer = tf.keras.layers.Dense(target_vocab_size)

    def call(self, inp, tar, training, enc_padding_mask, look_ahead_mask, dec_padding_mask):
        enc_output = self.encoder(inp, training=training, mask=enc_padding_mask)  # (batch_size, inp_seq_len, d_model)

        # dec_output.shape == (batch_size, tar_seq_len, d_model)
        dec_output, attention_weights = self.decoder(
            tar, enc_output, training=training, look_ahead_mask=look_ahead_mask, padding_mask=dec_padding_mask)

        final_output = self.final_layer(dec_output)  # (batch_size, tar_seq_len, target_vocab_size)

        return final_output, attention_weights

print("Transformer class updated with call method to pass arguments as keywords.")

Transformer class updated with call method to pass arguments as keywords.


**Reasoning**:
After updating the `Transformer` class's `call` method to correctly pass `training` and mask arguments as keyword arguments, I need to re-test the complete `Transformer` model by creating an instance with appropriate parameters and passing dummy input and target sequences to its `call` method. This will verify the output shape of the logits and the attention weights and confirm that the previous error has been resolved.



In [27]:
num_layers = 2
d_model = 512
num_heads = 8
dff = 2048

# Use the dummy vocab sizes and max sequence lengths from previous data preparation
input_vocab_size = len(source_vocab)
target_vocab_size = len(target_vocab)
pe_input = max_input_seq_len + 10 # Add some buffer for positional encoding
pe_target = max_target_seq_len + 10 # Add some buffer for positional encoding

# Create an instance of the Transformer model
transformer = Transformer(
    num_layers,
    d_model,
    num_heads,
    dff,
    input_vocab_size,
    target_vocab_size,
    pe_input,
    pe_target,
    rate=0.1
)

# Prepare dummy input and target sequences (using previously created batches)
# input_batch shape: (batch_size, inp_seq_len)
# target_batch shape: (batch_size, tar_seq_len)
dummy_input = input_batch # (3, 5)
dummy_target = target_batch[:, :-1] # Decoder input needs to be shifted right (remove last token)

# Generate masks
enc_padding_mask = create_padding_mask(dummy_input)
dec_padding_mask = create_padding_mask(dummy_input) # Used for encoder-decoder attention
look_ahead_mask = create_look_ahead_mask(tf.shape(dummy_target)[1])
combined_mask = tf.maximum(create_padding_mask(dummy_target), look_ahead_mask) # For decoder self-attention

# Call the transformer model
# Note: training=True to enable dropout
final_output, attention_weights = transformer(
    dummy_input,
    dummy_target,
    training=True,
    enc_padding_mask=enc_padding_mask,
    look_ahead_mask=combined_mask,
    dec_padding_mask=dec_padding_mask
)

print(f"Transformer output (logits) shape: {final_output.shape}") # Expected: (batch_size, tar_seq_len, target_vocab_size)
# Example of one attention weight shape from the decoder
if 'decoder_layer1_block1' in attention_weights:
    print(f"Sample attention weights shape (decoder_layer1_block1): {attention_weights['decoder_layer1_block1'].shape}")

TypeError: Exception encountered when calling Encoder.call().

'PositionalEncoding' object is not subscriptable

Arguments received by Encoder.call():
  • x=tf.Tensor(shape=(3, 5), dtype=int32)
  • training=True
  • mask=tf.Tensor(shape=(3, 1, 1, 5), dtype=float32)

**Reasoning**:
The previous error 'PositionalEncoding' object is not subscriptable occurred because the `Encoder` and `Decoder` classes were attempting to subscript the `PositionalEncoding` layer object itself (`self.pos_encoding[:, :seq_len, :]`) instead of calling its `call` method (`self.pos_encoding(x)`). I will modify both the `Encoder` and `Decoder` classes' `call` methods to correctly apply the positional encoding layer.



In [28]:
class Encoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Encoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(input_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, self.d_model)

        self.enc_layers = [
            EncoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]

        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, training, mask):
        # x.shape == (batch_size, input_seq_len)

        x = self.embedding(x)  # (batch_size, input_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        # Apply positional encoding by calling the layer
        x = self.pos_encoding(x) # Corrected line

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x = self.enc_layers[i](x, training, mask)

        return x  # (batch_size, input_seq_len, d_model)

print("Encoder class updated with corrected positional encoding application.")

Encoder class updated with corrected positional encoding application.


**Reasoning**:
As stated in the previous reasoning, the `Decoder` class also needs to be modified to correctly apply the positional encoding layer by calling its `call` method, similar to the `Encoder` class fix. This ensures consistent and correct usage of the `PositionalEncoding` layer across the Transformer model.



In [29]:
class Decoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, target_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Decoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(target_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, d_model)

        self.dec_layers = [
            DecoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]
        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, enc_output, training, look_ahead_mask, padding_mask):
        # x.shape == (batch_size, target_seq_len)
        attention_weights = {}

        x = self.embedding(x)  # (batch_size, target_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        # Apply positional encoding by calling the layer
        x = self.pos_encoding(x) # Corrected line

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x, block1, block2 = self.dec_layers[i](x, enc_output, training,
                                                    look_ahead_mask, padding_mask)

            attention_weights[f'decoder_layer{i+1}_block1'] = block1
            attention_weights[f'decoder_layer{i+1}_block2'] = block2

        # x.shape == (batch_size, target_seq_len, d_model)
        return x, attention_weights

print("Decoder class updated with corrected positional encoding application.")

Decoder class updated with corrected positional encoding application.


**Reasoning**:
Now that both the `Encoder` and `Decoder` classes have been updated to correctly apply the positional encoding, I will re-test the complete `Transformer` model with dummy data and masks. This final test will verify that all previous errors have been resolved and that the model produces outputs with the expected shapes, thus completing the subtask.



In [30]:
num_layers = 2
d_model = 512
num_heads = 8
dff = 2048

# Use the dummy vocab sizes and max sequence lengths from previous data preparation
input_vocab_size = len(source_vocab)
target_vocab_size = len(target_vocab)
pe_input = max_input_seq_len + 10 # Add some buffer for positional encoding
pe_target = max_target_seq_len + 10 # Add some buffer for positional encoding

# Create an instance of the Transformer model
transformer = Transformer(
    num_layers,
    d_model,
    num_heads,
    dff,
    input_vocab_size,
    target_vocab_size,
    pe_input,
    pe_target,
    rate=0.1
)

# Prepare dummy input and target sequences (using previously created batches)
# input_batch shape: (batch_size, inp_seq_len)
# target_batch shape: (batch_size, tar_seq_len)
dummy_input = input_batch # (3, 5)
dummy_target = target_batch[:, :-1] # Decoder input needs to be shifted right (remove last token)

# Generate masks
enc_padding_mask = create_padding_mask(dummy_input)
dec_padding_mask = create_padding_mask(dummy_input) # Used for encoder-decoder attention
look_ahead_mask = create_look_ahead_mask(tf.shape(dummy_target)[1])
combined_mask = tf.maximum(create_padding_mask(dummy_target), look_ahead_mask) # For decoder self-attention

# Call the transformer model
# Note: training=True to enable dropout
final_output, attention_weights = transformer(
    dummy_input,
    dummy_target,
    training=True,
    enc_padding_mask=enc_padding_mask,
    look_ahead_mask=combined_mask,
    dec_padding_mask=dec_padding_mask
)

print(f"Transformer output (logits) shape: {final_output.shape}") # Expected: (batch_size, tar_seq_len, target_vocab_size)
# Example of one attention weight shape from the decoder
if 'decoder_layer1_block1' in attention_weights:
    print(f"Sample attention weights shape (decoder_layer1_block1): {attention_weights['decoder_layer1_block1'].shape}")

ValueError: Exception encountered when calling Encoder.call().

Only input tensors may be passed as positional arguments. The following argument value should be passed as a keyword argument: True (of type <class 'bool'>)

Arguments received by Encoder.call():
  • x=tf.Tensor(shape=(3, 5), dtype=int32)
  • training=True
  • mask=tf.Tensor(shape=(3, 1, 1, 5), dtype=float32)

**Reasoning**:
The previous error in `Encoder.call()` stemmed from passing `training` and `mask` positionally to `EncoderLayer` within the loop. I need to modify the `Encoder` class's `call` method to ensure these arguments are passed as keyword arguments to each `EncoderLayer` instance.



In [31]:
class Encoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, input_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Encoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(input_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, self.d_model)

        self.enc_layers = [
            EncoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]

        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, training, mask):
        # x.shape == (batch_size, input_seq_len)

        x = self.embedding(x)  # (batch_size, input_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        # Apply positional encoding by calling the layer
        x = self.pos_encoding(x) # Corrected line

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x = self.enc_layers[i](x, training=training, mask=mask) # Corrected line: pass as keyword arguments

        return x  # (batch_size, input_seq_len, d_model)

print("Encoder class updated with corrected positional encoding and keyword argument passing.")

Encoder class updated with corrected positional encoding and keyword argument passing.


**Reasoning**:
The `Encoder` class has been updated to correctly pass `training` and `mask` as keyword arguments to its internal `EncoderLayer` instances. However, the `Decoder` class also has a similar loop for `DecoderLayer` instances, and it's highly probable that it might encounter the same positional argument error if not fixed. Therefore, I need to update the `Decoder` class to ensure `training`, `look_ahead_mask`, and `padding_mask` are passed as keyword arguments to each `DecoderLayer` instance.



In [32]:
class Decoder(tf.keras.layers.Layer):
    def __init__(self, num_layers, d_model, num_heads, dff, target_vocab_size,
                 maximum_position_encoding, rate=0.1):
        super(Decoder, self).__init__()

        self.d_model = d_model
        self.num_layers = num_layers

        self.embedding = tf.keras.layers.Embedding(target_vocab_size, d_model)
        self.pos_encoding = PositionalEncoding(maximum_position_encoding, d_model)

        self.dec_layers = [
            DecoderLayer(d_model, num_heads, dff, rate)
            for _ in range(num_layers)
        ]
        self.dropout = tf.keras.layers.Dropout(rate)

    def call(self, x, enc_output, training, look_ahead_mask, padding_mask):
        # x.shape == (batch_size, target_seq_len)
        attention_weights = {}

        x = self.embedding(x)  # (batch_size, target_seq_len, d_model)
        x *= tf.math.sqrt(tf.cast(self.d_model, tf.float32))
        # Apply positional encoding by calling the layer
        x = self.pos_encoding(x) # Corrected line

        x = self.dropout(x, training=training)

        for i in range(self.num_layers):
            x, block1, block2 = self.dec_layers[i](x, enc_output, training=training,
                                                    look_ahead_mask=look_ahead_mask, padding_mask=padding_mask) # Corrected line: pass as keyword arguments

            attention_weights[f'decoder_layer{i+1}_block1'] = block1
            attention_weights[f'decoder_layer{i+1}_block2'] = block2

        # x.shape == (batch_size, target_seq_len, d_model)
        return x, attention_weights

print("Decoder class updated with corrected positional encoding and keyword argument passing.")

Decoder class updated with corrected positional encoding and keyword argument passing.


**Reasoning**:
Now that both the `Encoder` and `Decoder` classes have been updated to correctly pass the positional encoding, I will re-test the complete `Transformer` model with dummy data and masks. This final test will verify that all previous errors have been resolved and that the model produces outputs with the expected shapes, thus completing the subtask.



In [33]:
num_layers = 2
d_model = 512
num_heads = 8
dff = 2048

# Use the dummy vocab sizes and max sequence lengths from previous data preparation
input_vocab_size = len(source_vocab)
target_vocab_size = len(target_vocab)
pe_input = max_input_seq_len + 10 # Add some buffer for positional encoding
pe_target = max_target_seq_len + 10 # Add some buffer for positional encoding

# Create an instance of the Transformer model
transformer = Transformer(
    num_layers,
    d_model,
    num_heads,
    dff,
    input_vocab_size,
    target_vocab_size,
    pe_input,
    pe_target,
    rate=0.1
)

# Prepare dummy input and target sequences (using previously created batches)
# input_batch shape: (batch_size, inp_seq_len)
# target_batch shape: (batch_size, tar_seq_len)
dummy_input = input_batch # (3, 5)
dummy_target = target_batch[:, :-1] # Decoder input needs to be shifted right (remove last token)

# Generate masks
enc_padding_mask = create_padding_mask(dummy_input)
dec_padding_mask = create_padding_mask(dummy_input) # Used for encoder-decoder attention
look_ahead_mask = create_look_ahead_mask(tf.shape(dummy_target)[1])
combined_mask = tf.maximum(create_padding_mask(dummy_target), look_ahead_mask) # For decoder self-attention

# Call the transformer model
# Note: training=True to enable dropout
final_output, attention_weights = transformer(
    dummy_input,
    dummy_target,
    training=True,
    enc_padding_mask=enc_padding_mask,
    look_ahead_mask=combined_mask,
    dec_padding_mask=dec_padding_mask
)

print(f"Transformer output (logits) shape: {final_output.shape}") # Expected: (batch_size, tar_seq_len, target_vocab_size)
# Example of one attention weight shape from the decoder
if 'decoder_layer1_block1' in attention_weights:
    print(f"Sample attention weights shape (decoder_layer1_block1): {attention_weights['decoder_layer1_block1'].shape}")

Transformer output (logits) shape: (3, 6, 11)
Sample attention weights shape (decoder_layer1_block1): (3, 8, 6, 6)


## Define Custom Learning Rate Schedule and Optimizer

### Subtask:
Transformers often benefit from a specific learning rate schedule with a warm-up phase. Define this schedule and use an optimizer like Adam.


**Reasoning**:
As instructed, I will define the `CustomSchedule` class, inheriting from `tf.keras.optimizers.schedules.LearningRateSchedule`, and implement its `__init__` and `__call__` methods to compute the learning rate with a warm-up phase.



In [34]:
class CustomSchedule(tf.keras.optimizers.schedules.LearningRateSchedule):
    def __init__(self, d_model, warmup_steps=4000):
        super(CustomSchedule, self).__init__()

        self.d_model = tf.cast(d_model, tf.float32)
        self.warmup_steps = tf.cast(warmup_steps, tf.float32)

    def __call__(self, step):
        step = tf.cast(step, tf.float32)
        arg1 = tf.math.rsqrt(step)
        arg2 = step * (self.warmup_steps ** -1.5)

        return tf.math.rsqrt(self.d_model) * tf.math.minimum(arg1, arg2)

print("CustomSchedule class defined.")

CustomSchedule class defined.


**Reasoning**:
As instructed, I will create an instance of the `CustomSchedule` and then initialize the `Adam` optimizer with this custom learning rate schedule and specified parameters.



In [35]:
learning_rate = CustomSchedule(d_model, warmup_steps=4000)

optimizer = tf.keras.optimizers.Adam(
    learning_rate,
    beta_1=0.9,
    beta_2=0.98,
    epsilon=1e-9
)

print("Custom learning rate schedule and Adam optimizer initialized.")

Custom learning rate schedule and Adam optimizer initialized.


## Train the Transformer Model

### Subtask:
Train the complete transformer model using the prepared dataset, custom learning rate schedule, and an appropriate loss function (e.g., SparseCategoricalCrossentropy for sequence-to-sequence tasks).


**Reasoning**:
As instructed, the first step is to define the loss function using `tf.keras.losses.SparseCategoricalCrossentropy` with `from_logits=True` and define a custom accuracy metric.



In [36]:
loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
    from_logits=True, reduction='none')

def loss_function(real, pred):
    # Ensure we are not calculating loss for padding tokens
    mask = tf.math.logical_not(tf.math.equal(real, PAD_ID))
    loss_ = loss_object(real, pred)

    mask = tf.cast(mask, dtype=loss_.dtype)
    loss_ *= mask

    return tf.reduce_sum(loss_)/tf.reduce_sum(mask)

def accuracy_function(real, pred):
    accuracies = tf.equal(real, tf.argmax(pred, axis=2))

    mask = tf.math.logical_not(tf.math.equal(real, PAD_ID))
    accuracies = tf.math.logical_and(mask, accuracies)

    accuracies = tf.cast(accuracies, dtype=tf.float32)
    mask = tf.cast(mask, dtype=tf.float32)
    return tf.reduce_sum(accuracies)/tf.reduce_sum(mask)

print("Loss function and accuracy metric defined.")

Loss function and accuracy metric defined.


**Reasoning**:
Now that the loss function and accuracy metric are defined, the next step is to compile the `Transformer` model with the custom optimizer, loss, and accuracy functions, and then train it using the prepared dummy data, ensuring correct masking and shifted target labels for the decoder.



In [37]:
epochs = 5 # You can adjust this for more training

# Prepare the dataset for training
# The decoder input 'tar' is target_batch[:, :-1]
# The true labels 'tar_real' for loss calculation are target_batch[:, 1:]
tar_inp = dummy_target # This is already target_batch[:, :-1]
tar_real = target_batch[:, 1:] # Shifted target for calculating loss

# Verify shapes
print(f"Shape of encoder input (inp): {dummy_input.shape}")
print(f"Shape of decoder input (tar_inp): {tar_inp.shape}")
print(f"Shape of real target (tar_real): {tar_real.shape}")


# Training loop (simplified for demonstration, typically would use tf.data.Dataset)
for epoch in range(epochs):
    print(f"\nEpoch {epoch + 1}/{epochs}")

    # Generate masks for the current training step
    enc_padding_mask = create_padding_mask(dummy_input)
    dec_padding_mask = create_padding_mask(dummy_input) # Mask for encoder-decoder attention
    look_ahead_mask = create_look_ahead_mask(tf.shape(tar_inp)[1])
    # Combined mask for decoder self-attention: padding mask + look ahead mask
    combined_mask = tf.maximum(create_padding_mask(tar_inp), look_ahead_mask)

    with tf.GradientTape() as tape:
        predictions, _ = transformer(
            dummy_input,
            tar_inp,
            training=True,
            enc_padding_mask=enc_padding_mask,
            look_ahead_mask=combined_mask,
            dec_padding_mask=dec_padding_mask
        )

        current_loss = loss_function(tar_real, predictions)
        current_accuracy = accuracy_function(tar_real, predictions)

    gradients = tape.gradient(current_loss, transformer.trainable_variables)
    optimizer.apply_gradients(zip(gradients, transformer.trainable_variables))

    print(f"Loss: {current_loss.numpy():.4f}, Accuracy: {current_accuracy.numpy():.4f}")

print("Transformer model training complete.")

Shape of encoder input (inp): (3, 5)
Shape of decoder input (tar_inp): (3, 6)
Shape of real target (tar_real): (3, 6)

Epoch 1/5


InvalidArgumentError: cannot compute Equal as input #1(zero-based) was expected to be a int32 tensor but is a int64 tensor [Op:Equal] name: 

**Reasoning**:
The `InvalidArgumentError` in the `accuracy_function` indicates a type mismatch between `real` (int32) and the output of `tf.argmax(pred, axis=2)` (int64). I need to cast the output of `tf.argmax` to `tf.int32` to resolve this.



In [38]:
epochs = 5 # You can adjust this for more training

# Prepare the dataset for training
# The decoder input 'tar' is target_batch[:, :-1]
# The true labels 'tar_real' for loss calculation are target_batch[:, 1:]
tar_inp = dummy_target # This is already target_batch[:, :-1]
tar_real = target_batch[:, 1:] # Shifted target for calculating loss

# Verify shapes
print(f"Shape of encoder input (inp): {dummy_input.shape}")
print(f"Shape of decoder input (tar_inp): {tar_inp.shape}")
print(f"Shape of real target (tar_real): {tar_real.shape}")

def accuracy_function(real, pred):
    # Cast the argmax output to int32 to match 'real' tensor's dtype
    accuracies = tf.equal(real, tf.cast(tf.argmax(pred, axis=2), tf.int32))

    mask = tf.math.logical_not(tf.math.equal(real, PAD_ID))
    accuracies = tf.math.logical_and(mask, accuracies)

    accuracies = tf.cast(accuracies, dtype=tf.float32)
    mask = tf.cast(mask, dtype=tf.float32)
    return tf.reduce_sum(accuracies)/tf.reduce_sum(mask)


# Training loop (simplified for demonstration, typically would use tf.data.Dataset)
for epoch in range(epochs):
    print(f"\nEpoch {epoch + 1}/{epochs}")

    # Generate masks for the current training step
    enc_padding_mask = create_padding_mask(dummy_input)
    dec_padding_mask = create_padding_mask(dummy_input) # Mask for encoder-decoder attention
    look_ahead_mask = create_look_ahead_mask(tf.shape(tar_inp)[1])
    # Combined mask for decoder self-attention: padding mask + look ahead mask
    combined_mask = tf.maximum(create_padding_mask(tar_inp), look_ahead_mask)

    with tf.GradientTape() as tape:
        predictions, _ = transformer(
            dummy_input,
            tar_inp,
            training=True,
            enc_padding_mask=enc_padding_mask,
            look_ahead_mask=combined_mask,
            dec_padding_mask=dec_padding_mask
        )

        current_loss = loss_function(tar_real, predictions)
        current_accuracy = accuracy_function(tar_real, predictions)

    gradients = tape.gradient(current_loss, transformer.trainable_variables)
    optimizer.apply_gradients(zip(gradients, transformer.trainable_variables))

    print(f"Loss: {current_loss.numpy():.4f}, Accuracy: {current_accuracy.numpy():.4f}")

print("Transformer model training complete.")

Shape of encoder input (inp): (3, 5)
Shape of decoder input (tar_inp): (3, 6)
Shape of real target (tar_real): (3, 6)

Epoch 1/5
Loss: 3.3204, Accuracy: 0.1333

Epoch 2/5
Loss: 3.2240, Accuracy: 0.1333

Epoch 3/5
Loss: 3.0856, Accuracy: 0.2000

Epoch 4/5
Loss: 3.1914, Accuracy: 0.1333

Epoch 5/5
Loss: 3.1485, Accuracy: 0.1333
Transformer model training complete.


## Evaluate the Transformer Model

### Subtask:
Evaluate the trained transformer model on a test dataset, possibly using metrics like BLEU score for machine translation tasks, or accuracy for other sequence tasks.


**Reasoning**:
As instructed, the first step is to set up a function for making predictions with the Transformer model. This function will iteratively generate tokens, handling the necessary masking and utilizing the previously defined `transformer` model components.



In [39]:
def evaluate(inp_sentence):
    encoder_input = tf.expand_dims(inp_sentence, 0)

    # The decoder input is initially just the start token
    decoder_input = tf.expand_dims([target_id_to_word['<start>']], 0)

    output = tf.TensorArray(tf.int32, size=0, dynamic_size=True)

    # Assuming max_target_seq_len is defined and accessible
    # If not, we would need to pass it or infer it
    for i in range(max_target_seq_len): # Iterate up to max_target_seq_len
        enc_padding_mask = create_padding_mask(encoder_input)
        dec_padding_mask = create_padding_mask(encoder_input) # Mask for encoder-decoder attention

        look_ahead_mask = create_look_ahead_mask(tf.shape(decoder_input)[1])
        # Combined mask for decoder self-attention: padding mask + look ahead mask
        combined_mask = tf.maximum(create_padding_mask(decoder_input), look_ahead_mask)

        # predictions.shape == (batch_size, seq_len, vocab_size)
        predictions, attention_weights = transformer(
            encoder_input,
            decoder_input,
            training=False, # Set to False for evaluation mode
            enc_padding_mask=enc_padding_mask,
            look_ahead_mask=combined_mask,
            dec_padding_mask=dec_padding_mask
        )

        # Select the last token from the seq_len dimension
        predictions = predictions[:, -1:, :]

        predicted_id = tf.cast(tf.argmax(predictions, axis=-1), tf.int32)

        # Return the prediction if it's the end token
        if tf.equal(predicted_id, target_id_to_word['<end>']):
            break

        # Concatentate the predicted_id to the output which is given to the decoder
        # as its input.
        decoder_input = tf.concat([decoder_input, predicted_id], axis=-1)
        output = output.write(i, predicted_id[0, 0])

    return tf.expand_dims(output.stack(), 0), attention_weights

print("Prediction function 'evaluate' defined.")

Prediction function 'evaluate' defined.


**Reasoning**:
As instructed, the next step is to apply the `evaluate` function to the dummy input sequences to get predictions, then compare these predictions with the actual target sequences using the `accuracy_function`, and finally print the resulting accuracy to assess the model's performance.



In [40]:
predicted_sequences = []
actual_target_sequences = []

# Iterate through each sample in the dummy_input batch
for i in range(tf.shape(dummy_input)[0]):
    inp_sample = dummy_input[i]
    actual_target_sample = tar_real[i] # Get the corresponding actual target sequence

    # Generate prediction for the current input sample
    result, _ = evaluate(inp_sample)

    # result.shape is (1, sequence_length), remove the batch dimension
    predicted_sequences.append(tf.squeeze(result))
    actual_target_sequences.append(actual_target_sample)

# Pad predicted sequences to match the length of tar_real for accuracy calculation
# The padding token should be PAD_ID (0)

# Find the maximum length among predicted sequences for padding
max_predicted_len = max(tf.shape(s)[0] for s in predicted_sequences)

# Pad each predicted sequence
final_predicted_sequences = []
for seq in predicted_sequences:
    current_len = tf.shape(seq)[0]
    padding_needed = max_predicted_len - current_len
    if padding_needed > 0:
        # Pad with PAD_ID (0) at the end
        padded_seq = tf.pad(seq, [[0, padding_needed]], "CONSTANT", constant_values=PAD_ID)
    else:
        padded_seq = seq
    final_predicted_sequences.append(padded_seq)

# Convert lists of tensors to a single tensor for batch processing
# Need to ensure all sequences have the same length after padding for stacking
final_predicted_sequences_tensor = tf.stack(final_predicted_sequences)
actual_target_sequences_tensor = tf.stack(actual_target_sequences)

# Ensure the lengths match for accuracy calculation by potentially truncating or further padding
# If max_predicted_len > actual_target_sequences_tensor.shape[1], we might need to truncate predictions
# If max_predicted_len < actual_target_sequences_tensor.shape[1], we need to pad predictions or handle mismatch

# For simplicity, let's assume tar_real's length is the reference and truncate/pad predictions
reference_len = tf.shape(tar_real)[1]

# Adjust final_predicted_sequences_tensor to match reference_len
if tf.shape(final_predicted_sequences_tensor)[1] > reference_len:
    final_predicted_sequences_tensor = final_predicted_sequences_tensor[:, :reference_len]
elif tf.shape(final_predicted_sequences_tensor)[1] < reference_len:
    padding_needed_final = reference_len - tf.shape(final_predicted_sequences_tensor)[1]
    final_predicted_sequences_tensor = tf.pad(final_predicted_sequences_tensor, [[0, 0], [0, padding_needed_final]], "CONSTANT", constant_values=PAD_ID)


# Reshape predictions to match the (batch_size, seq_len, vocab_size) format expected by accuracy_function
# We'll create a one-hot like representation for the predicted IDs for accuracy_function

batch_size = tf.shape(final_predicted_sequences_tensor)[0]
seq_len_pred = tf.shape(final_predicted_sequences_tensor)[1]
vocab_size = target_vocab_size

# Create a dummy one-hot tensor where the predicted ID has a high logit
# This is a simplification; in a real scenario, you'd use the model's logits
# For this demonstration, we are just comparing predicted_id with real_id.
# The accuracy_function actually expects logits for `pred` argument.
# So, we need to create a tensor that has the correct shape and can be argmaxed.

# Create a tensor of zeros with shape (batch_size, seq_len_pred, vocab_size)
# Set the value at the predicted index to 1 (or any high value) for argmax to pick it.
predicted_logits_for_accuracy = tf.one_hot(final_predicted_sequences_tensor, depth=vocab_size, dtype=tf.float32) * 100.0

# Calculate accuracy using the defined accuracy_function
# Use tar_real as the 'real' values for comparison
overall_accuracy = accuracy_function(tar_real, predicted_logits_for_accuracy)

print(f"Predicted sequence shape (after padding/truncation): {final_predicted_sequences_tensor.shape}")
print(f"Actual target sequence shape: {tar_real.shape}")
print(f"Overall Evaluation Accuracy: {overall_accuracy.numpy():.4f}")

# Optionally, print a sample prediction for verification
print("\nSample Prediction:")
for i in range(min(len(input_sequences_raw), 1)): # Print the first sample
    input_decoded = decode_sequence(input_batch[i].numpy(), source_vocab)
    actual_decoded = decode_sequence(tar_real[i].numpy(), target_vocab)
    predicted_decoded = decode_sequence(final_predicted_sequences_tensor[i].numpy(), target_vocab)

    print(f"Input: {input_decoded}")
    print(f"Actual: {actual_decoded}")
    print(f"Predicted: {predicted_decoded}")

Predicted sequence shape (after padding/truncation): (3, 6)
Actual target sequence shape: (3, 6)
Overall Evaluation Accuracy: 0.1333

Sample Prediction:
Input: hello world how are you
Actual: bonjour monde comment allez vous <end>
Predicted: allez allez allez allez allez allez


## Final Task

### Subtask:
Summarize the implementation process and the performance of the transformer model.


## Summary:

### Q&A
The task was to summarize the implementation process and the performance of the transformer model.

*   **Implementation Process**: The implementation followed a modular approach, starting with an understanding of Transformer architecture, then building individual components such as Positional Encoding, Multi-Head Attention, Encoder Layer, and Decoder Layer. These components were then assembled into a full Transformer model. Custom masks, a learning rate schedule, and an optimizer were defined. Finally, a training loop and an evaluation function were implemented. The process involved several debugging steps, particularly during the full model assembly and training, to resolve issues related to Keras layer argument passing and data type mismatches.
*   **Performance of the Transformer Model**: The model's performance on the dummy dataset was very poor, achieving an overall evaluation accuracy of approximately $0.1333$. Qualitative observation showed that the model generated highly repetitive predictions (e.g., "allez allez allez allez allez allez") instead of meaningful translations.

### Data Analysis Key Findings
*   **Transformer Architecture**: A comprehensive understanding of key components like self-attention, multi-head attention, positional encoding, and the encoder-decoder structure was established.
*   **Positional Encoding**: The `PositionalEncoding` class was successfully implemented, generating sinusoidal encodings of shape `(1, 50, 512)` for a `max_sequence_length` of 50 and `d_model` of 512.
*   **Data Preparation**: Dummy source and target vocabularies were defined, and raw numericalized sequences were created. These were then padded to `(3, 5)` for input and `(3, 7)` for target, and batched into TensorFlow tensors, including `<start>`, `<end>`, and `<pad>` tokens.
*   **Multi-Head Attention**: A `MultiHeadAttention` layer and the `scaled_dot_product_attention` function were implemented. Testing with dummy inputs (`batch_size=64`, `seq_len=100`, `d_model=512`, `num_heads=8`) resulted in an output shape of `(64, 100, 512)` and attention weights of `(64, 8, 100, 100)`, confirming correct functionality.
*   **Encoder and Decoder Layers**: A single `EncoderLayer` and `DecoderLayer` were successfully implemented, incorporating multi-head attention, feed-forward networks, residual connections, layer normalization, and dropout.
*   **Full Transformer Model Assembly**: The complete `Transformer` model was assembled, integrating the `Encoder` and `Decoder` components. This involved defining `create_padding_mask` and `create_look_ahead_mask` functions. During assembly and testing, several critical errors were identified and resolved, including:
    *   Incorrect positional argument passing to Keras layers within `Transformer.call`.
    *   Incorrect application of `PositionalEncoding` (subscripting instead of calling the layer) within `Encoder.call` and `Decoder.call`.
    *   Incorrect positional argument passing to `EncoderLayer` and `DecoderLayer` instances.
    After resolution, the model produced correct output shapes: logits `(batch_size, tar_seq_len, target_vocab_size)` and attention weights `(batch_size, num_heads, tar_seq_len, tar_seq_len)`.
*   **Learning Rate Schedule and Optimizer**: A `CustomSchedule` learning rate policy was implemented, featuring a warm-up phase, and used to initialize an `Adam` optimizer with `beta_1=0.9`, `beta_2=0.98`, and `epsilon=1e-9`.
*   **Training Process**: A training loop was implemented with `tf.keras.losses.SparseCategoricalCrossentropy` and a custom `accuracy_function` that masked padding tokens. During initial training, a `TypeError` was encountered in `accuracy_function` due to a mismatch between `real` (int32) and `tf.argmax(pred, axis=2)` (int64), which was resolved by casting `tf.argmax` output to `tf.int32`. The model was trained for 5 epochs on the dummy data.
*   **Evaluation**: An `evaluate` function was defined for inference. The model was evaluated on the dummy test set, resulting in a low overall accuracy of approximately $0.1333$. Qualitative analysis of sample predictions revealed repetitive output tokens.

### Insights or Next Steps
*   The current model performance is poor, likely due to training on a very small and simple dummy dataset with only 3 samples for 5 epochs. To achieve meaningful results, the model requires training on a much larger and more diverse dataset relevant to a specific NLP task (e.g., machine translation with millions of sentence pairs) and for significantly more epochs.
*   Further hyperparameter tuning (e.g., `d_model`, `num_heads`, `dff`, `dropout_rate`, `warmup_steps`, number of layers) and optimization techniques (e.g., gradient clipping, label smoothing) would be essential to improve model performance on a real-world task.
